presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/TestPushProjectionThroughUnion.java

```diff
@@ -0,0 +1,82 @@
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableListMultimap;
+import com.google.common.collect.ImmutableMap;
+import org.testng.annotations.Test;
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.expression;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.project;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.union;
+import static com.facebook.presto.sql.planner.assertions.PlanMatchPattern.values;
+
+public class TestPushProjectionThroughUnion
+{
+    private final RuleTester tester = new RuleTester();
+
+    @Test
+    public void testDoesNotFire()
+            throws Exception
+    {
+        tester.assertThat(new PushProjectionThroughUnion())
+                .on(p ->
+                        p.project(
+                                Assignments.of(p.symbol("x", BIGINT), new LongLiteral("3")),
+                                p.values(p.symbol("a", BIGINT))))
+                .doesNotFire();
+    }
+
+    @Test
+    public void test()
+            throws Exception
+    {
+        tester.assertThat(new PushProjectionThroughUnion())
+                .on(p -> {
+                    Symbol a = p.symbol("a", BIGINT);
+                    Symbol b = p.symbol("b", BIGINT);
+                    Symbol c = p.symbol("c", BIGINT);
+                    Symbol x = p.symbol("x", BIGINT);
+                    return p.project(
+                            Assignments.of(x, new LongLiteral("3")),
+                            p.union(
+                                    ImmutableList.of(
+                                            p.values(a),
+                                            p.values(b)),
+                                    ImmutableListMultimap.<Symbol, Symbol>builder()
+                                            .put(c, a)
+                                            .put(c, b)
+                                            .build(),
+                                    ImmutableList.of(c)));
+                })
+                .matches(
+                        union(
+                                project(
+                                        ImmutableMap.of("x1", expression("3")),
+                                        values(ImmutableMap.of())),
+                                project(
+                                        ImmutableMap.of("x2", expression("3")),
+                                        values(ImmutableMap.of())))
+                                // verify that data originally on symbols aliased as x1 and x2 is part of exchange output
+                                .withAlias("x1")
+                                .withAlias("x2"));
+        // todo check contents of values
```

Comment:
maybe you could please add support for `values`. I see this TODO in so many rule tests that I am surprised that nobody fixed that so far.